File: library/src/main/java/com/github/dmitrikudrenko/countryphonecodechooser/ui/adapter/sections/FilterCountryComparator.java
package com.github.dmitrikudrenko.countryphonecodechooser.ui.adapter.sections;

import android.support.annotation.RestrictTo;
import android.support.annotation.RestrictTo.Scope;
import com.github.dmitrikudrenko.countryphonecodechooser.model.CountryCode;

import java.util.Comparator;

@RestrictTo(Scope.LIBRARY)
public class FilterCountryComparator implements Comparator<SectionCountry> {
    private static final int MAX_PRIORITY = 10;
    private String filter;

    public FilterCountryComparator(final String filter) {
        this.filter = filter;
    }

    @Override
    public int compare(final SectionCountry firstSectionCountry,
                       final SectionCountry secondSectionCountry) {
        final CountryCode firstCode = firstSectionCountry.getCode();
        final CountryCode secondCode = secondSectionCountry.getCode();

        final int firstCodePriority = getEntryPriority(firstCode.getName());
        final int secondCodePriority = getEntryPriority(secondCode.getName());
        if (firstCodePriority > secondCodePriority) {
            return -1;
        } else if (secondCodePriority > firstCodePriority) {
            return 1;
        } else {
            return firstCode.compareTo(secondCode);
        }
    }

    private int getEntryPriority(final String value) {
        final String[] words = value.split(" ");
        for (int i = 0; i < words.length; i++) {
            final String word = words[i];
            if (word.toLowerCase().startsWith(filter.toLowerCase())) {
                return MAX_PRIORITY - i;
            }
        }
        return 1;
    }
}

File: library/src/main/kotlin/com/github/dmitrikudrenko/countryphonecodechooser/ui/adapter/sections/FilterCountryComparator.kt
package com.github.dmitrikudrenko.countryphonecodechooser.ui.adapter.sections

import android.support.annotation.RestrictTo
import android.support.annotation.RestrictTo.Scope
import java.util.*

private const val MAX_PRIORITY = 10

@RestrictTo(Scope.LIBRARY)
class FilterCountryComparator internal constructor(private val filter: String)
    : Comparator<SectionCountry> {

    override fun compare(firstSectionCountry: SectionCountry,
                         secondSectionCountry: SectionCountry): Int {
        val firstCode = firstSectionCountry.code
        val secondCode = secondSectionCountry.code

        val firstCodePriority = firstCode.name.let { getEntryPriority(it) }
        val secondCodePriority = secondCode.name.let { getEntryPriority(it) }

        return when {
            firstCodePriority > secondCodePriority -> -1
            secondCodePriority > firstCodePriority -> 1
            else -> firstCode.compareTo(secondCode)
        }
    }

    private fun getEntryPriority(value: String): Int {
        val words = value.split(" ".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
        for ((i, word) in words.withIndex()) {
            if (word.toLowerCase().startsWith(filter.toLowerCase())) {
                return MAX_PRIORITY - i
            }
        }
        return 1
    }
}
